Svc/GenericHub/GenericHubComponentImpl.cpp

```diff
@@ -96,13 +94,85 @@ void GenericHubComponentImpl ::dataIn_handler(const NATIVE_INT_TYPE portNum,
         status = wrapper.setBuffLen(rawSize);
         FW_ASSERT(status == Fw::FW_SERIALIZE_OK, static_cast<NATIVE_INT_TYPE>(status));
         portOut_out(port, wrapper);
+        // Deallocate the existing buffer
         dataInDeallocate_out(0, fwBuffer);
     } else if (type == HUB_TYPE_BUFFER) {
+        // Fw::Buffers can reuse the existing data buffer as the storage type!  No deallocation done.
         fwBuffer.set(rawData, rawSize, fwBuffer.getContext());
         buffersOut_out(port, fwBuffer);
+    } else if (type == HUB_TYPE_EVENT) {
+        FwEventIdType id;
+        Fw::Time timeTag;
+        Fw::LogSeverity severity;
+        Fw::LogBuffer args;
+
+        // Deserialize tokens for events
+        status = incoming.deserialize(id);
+        FW_ASSERT(status == Fw::FW_SERIALIZE_OK, static_cast<NATIVE_INT_TYPE>(status));
+        status = incoming.deserialize(timeTag);
+        FW_ASSERT(status == Fw::FW_SERIALIZE_OK, static_cast<NATIVE_INT_TYPE>(status));
+        status = incoming.deserialize(severity);
+        FW_ASSERT(status == Fw::FW_SERIALIZE_OK, static_cast<NATIVE_INT_TYPE>(status));
+        status = incoming.deserialize(args);
+        FW_ASSERT(status == Fw::FW_SERIALIZE_OK, static_cast<NATIVE_INT_TYPE>(status));
+
+        // Send it!
+        this->LogSend_out(port, id, timeTag, severity, args);
+
+        // Deallocate the existing buffer
+        dataInDeallocate_out(0, fwBuffer);
+    } else if (type == HUB_TYPE_CHANNEL) {
+        FwChanIdType id;
+        Fw::Time timeTag;
+        Fw::TlmBuffer val;
+
+        // Deserialize tokens for channels
+        status = incoming.deserialize(id);
+        FW_ASSERT(status == Fw::FW_SERIALIZE_OK, static_cast<NATIVE_INT_TYPE>(status));
+        status = incoming.deserialize(timeTag);
+        FW_ASSERT(status == Fw::FW_SERIALIZE_OK, static_cast<NATIVE_INT_TYPE>(status));
+        status = incoming.deserialize(val);
+        FW_ASSERT(status == Fw::FW_SERIALIZE_OK, static_cast<NATIVE_INT_TYPE>(status));
+
+        // Send it!
+        this->TlmSend_out(port, id, timeTag, val);
+
+        // Deallocate the existing buffer
+        dataInDeallocate_out(0, fwBuffer);
     }
 }
 
+void GenericHubComponentImpl ::LogRecv_handler(const NATIVE_INT_TYPE portNum,
+                                  FwEventIdType id,
+                                  Fw::Time& timeTag,
+                                  const Fw::LogSeverity& severity,
+                                  Fw::LogBuffer& args) {
+    U8 buffer[sizeof(FwEventIdType) + Fw::Time::SERIALIZED_SIZE + Fw::LogSeverity::SERIALIZED_SIZE + FW_LOG_BUFFER_MAX_SIZE];
+    Fw::ExternalSerializeBuffer serializer(buffer, sizeof(buffer));
+    serializer.resetSer();
+    serializer.serialize(id);
```

Comment:
Also, to be symmetric with the decode function, why aren't you serializing the port type (`HUB_TYPE_EVENT`,etc)

Reply:
I am, it it done in the `send_data` helper.  This level function is just for wrangling the specific data into a buffer that function could serialize.